cmd/kubeadm/app/util/config/joinconfiguration.go

```diff
@@ -60,54 +60,76 @@ func SetJoinControlPlaneDefaults(cfg *kubeadmapi.JoinControlPlane) error {
-
 	if cfgPath != "" {
 		// Loads configuration from config file, if provided
 		// Nb. --config overrides command line flags, TODO: fix this
-		klog.V(1).Infoln("loading configuration from the given file")
+		return LoadJoinConfigurationFromFile(cfgPath)
+	}
+
+	return DefaultedJoinConfiguration(defaultversionedcfg)
+}
 
-		b, err := ioutil.ReadFile(cfgPath)
-		if err != nil {
-			return nil, errors.Wrapf(err, "unable to read config from %q ", cfgPath)
+// LoadJoinConfigurationFromFile loads versioned JoinConfiguration from file, converts it to internal, defaults and validates it
+func LoadJoinConfigurationFromFile(cfgPath string) (*kubeadmapi.JoinConfiguration, error) {
+	klog.V(1).Infof("loading configuration from %q", cfgPath)
+
+	b, err := ioutil.ReadFile(cfgPath)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to read config from %q ", cfgPath)
+	}
+
+	gvkmap, err := kubeadmutil.SplitYAMLDocuments(b)
+	if err != nil {
+		return nil, err
+	}
+
+	joinBytes := []byte{}
+	for gvk, bytes := range gvkmap {
+		// not interested in anything other than JoinConfiguration
+		if gvk.Kind != constants.JoinConfigurationKind {
+			continue
 		}
 
-		gvkmap, err := kubeadmutil.SplitYAMLDocuments(b)
-		if err != nil {
+		// check if this version is supported one
+		if err := ValidateSupportedVersion(gvk.GroupVersion()); err != nil {
 			return nil, err
 		}
 
-		joinBytes := []byte{}
-		for gvk, bytes := range gvkmap {
-			// not interested in anything other than JoinConfiguration
-			if gvk.Kind != constants.JoinConfigurationKind {
-				continue
-			}
+		// verify the validity of the YAML
+		strict.VerifyUnmarshalStrict(bytes, gvk)
 
-			// check if this version is supported one
-			if err := ValidateSupportedVersion(gvk.GroupVersion()); err != nil {
-				return nil, err
-			}
-
-			// verify the validity of the YAML
-			strict.VerifyUnmarshalStrict(bytes, gvk)
+		joinBytes = bytes
+	}
 
-			joinBytes = bytes
-		}
+	if len(joinBytes) == 0 {
+		return nil, errors.Errorf("no %s found in config file %q", constants.JoinConfigurationKind, cfgPath)
+	}
 
-		if len(joinBytes) == 0 {
-			return nil, errors.Errorf("no %s found in config file %q", constants.JoinConfigurationKind, cfgPath)
-		}
+	internalcfg := &kubeadmapi.JoinConfiguration{}
+	if err := runtime.DecodeInto(kubeadmscheme.Codecs.UniversalDecoder(), joinBytes, internalcfg); err != nil {
+		return nil, err
+	}
 
-		if err := runtime.DecodeInto(kubeadmscheme.Codecs.UniversalDecoder(), joinBytes, internalcfg); err != nil {
-			return nil, err
-		}
-	} else {
-		// Takes passed flags into account; the defaulting is executed once again enforcing assignement of
-		// static default values to cfg only for values not provided with flags
-		kubeadmscheme.Scheme.Default(defaultversionedcfg)
-		kubeadmscheme.Scheme.Convert(defaultversionedcfg, internalcfg, nil)
+	// Applies dynamic defaults to settings not provided with flags
+	if err := SetJoinDynamicDefaults(internalcfg); err != nil {
```

Comment:
If I'm not wrong, we should apply defaults before, and then dynamic defaults (otherwise the user will be provided to set static default values manually)
Wdyt?

Reply:
The preceding call to `DecodeInto` applies static defaults upon decoding the object. That's why it actually works in the current code base.